MEMMODEL_CONSTRAINTS.md

```diff
@@ -103,6 +103,19 @@ may perceive in that statement.)
   for memory used for atomics.  The situation is somewhat better in
   well-typed C++ but reality is complicated.
 
+## Implementation targeting C and C++ or LLVM
+
+* _Targeting C or C++_: it should be possible to implement
+  SharedArrayBuffer using C or C++.  The C/C++11 memory model allows
+  accesses with a number of orderings, and there should be a mapping
+  from SharedArrayBuffer accesses to C/C++11 orderings. Racy
+  SharedArrayBuffer accesses are required to produce defined values,
+  so cannot be implemented using C/C++ non-atomics.
+* _Targeting LLVM_: implementations may also decide to target
+  LLVM, which may provide for more optimization than targeting C/C++
+  and compiling to LLVM. In particular, LLVM allows unordered access,
+  which does not require per-variable sequential consistency.
```

Comment:
>>> If we map SAB non-atomic accesses to LLVM's unordered, I think this all the same optimizations as LLVM non-atomic. Slightly annoyingly, C11's relaxed blocks some optimizations due to coherence.

>> unordered isn't the same as LLVM's non-atomics.

> Indeed, I'm asking whether we can implement SAB non-atomic as LLVM unordered? (In particular, is there a requirement that SAB non-atomics are coherent?)

What do you mean by "implement"?

Are you trying to compile asm.js / wasm code using SAB to LLVM IR, and trying to figure out which load / store type to use?

Or are you trying to figure out how to specify the semantics of SAB?

>> relaxed basically only enforces same-address writes are seen in a thread's write order. You can totally reorder them, fuse them, move them around, eliminate dead ones, etc, and there's no guarantee that a relaxed write will ever be observed without synchronization actions occurring.

> Relaxed operations can't be reordered as much as unordered operations, due to the requirement of coherence. For example, x=1;x=2;x=3 can't be optimized to x=2;x=1;x=3 for relaxed writes, but can for unordered.

Sure, but that doesn't make my statement any less true:
* Both can be optimized to x=3.
* x+=1;y+=2;x+=3;y+=4; can be optimized to x+=4;y+=6; in both cases.

>>>    Targeting C or C++: it should be possible to implement
>>>        SharedArrayBuffer using C or C++. The C/C++11 memory model allows

>> The wording is confusing: C/C++ should be mappable to SAB. Or are you saying that the implementation of SAB (e.g. a JS engine) is in C++? That's not really true either, because the SAB code is JIT-compiled.

> I'm meaning the implementation of SAB (e.g. inside a JS engine). The question is, should it be possible to implement SAB in C++ or LLVM, or does the implementation have to be in hand-crafted asm? (This impacts my life, as someone who may be implementing the Servo interactions with SAB, if they end up being visible in the browser.)

Implementing SAB -> LLVM IR -> x86/ARM/etc should totally be possible. As you point out it's not clear what restrictions the SAB spec is currently putting on what LLVM IR you can generate.

I'll go further to say: that shouldn't be a significant design criteria. It should be possible to lower SAB to efficient machine code, but whether LLVM is well suited or not is a fixable issue.

>>>    accesses with a number of orderings, and there should be a mapping
>>>    from SharedArrayBuffer accesses to C/C++11 orderings. Racy

>> That's currently not true: only seq_cst is supported by SAB. There are issues open for other ones.

> OK, what word should I use for non-atomic accesses to an SAB? Presumably non-atomic access is being mapped to something weaker than seq_cst.

Now that I understand what you're trying to do (SAB -> LLVM IR -> ...) I think mentioning C++ is misleading.

Yes, SAB's non-atomics are weaker than `seq_cst`. It's still not well defined what that is, and exactly how it interacts with atomics.

>>>    SharedArrayBuffer accesses are required to produce defined values,
>>>    so cannot be implemented using C/C++ non-atomics.

>> That's not really accurate: they simply don't have the same semantics. C++ has much looser guarantees, both when it comes to races (they're simply UB) and alignment. As above I don't see SAB as being "implemented using C++".

> Yes,. that's the point I'm trying to make here, SAB non-atomic accesses can't be mapped to C11 non-atomics, since racy non-atomics are UB in C11.

That's not quite right: an implementation ca decide to specify the UB we care about here. I think the question you're trying to get to is: does LLVM implement SAB's memory model so your job is simpler. I don't know because:

* SAB's memory model isn't finalized (`seq_cst` is easy, non-atomics and their interactions aren't yet).
* LLVM's optimization of atomics keep improving (maybe it does what you want now, but that may change in the future, because LLVM IR handles atomics for C++ and Java, not for SAB).

>>> +* Targeting LLVM: implementations may also decide to target
>>>
>>>    the LLVM memory model, for example by using a language such as
>>>    Rust, which inherits its memory model from LLVM. LLVM has a very
>>>    similar model to C/C++, but allows mixed-size access, and
>>>    allows Java-like unordered access.

>> I still don't think this is accurate. Implementations of what? Are we talking about Rust -> LLVM -> SAB, or about ??? -> SAB -> LLVM?

> ??? -> SAB -> LLVM. This whole section is about how SAB can be implemented, not how to use it as a compilation target.

That wasn't clear at all to me. I'd appreciate it if you could clarify the wording.

>> I'd also like to clarify what you mean by Rust. Sure it uses LLVM, but I don't think it's accurate to say that its memory model is LLVM's.

> OK, I talked to @nikomatsakis, and the Rust memory model is still a work in progress, so while the current implementation is via LLVM (and so inherits its memory model) future implementations may provide weaker guarantees.

I wouldn't mention it if it's a moving target. Further, if Rust's recommended parallelism datastructures are shared-nothing then this discussion is kinda misleading. I'd at least like a link referencing the Rust memory model you have in mind.